qutip/tests/conftest.py

```diff
@@ -0,0 +1,197 @@
+#
+#    Copyright (c) 2011 and later, Paul D. Nation and Robert J. Johansson.
+#    All rights reserved.
+#
+#    Redistribution and use in source and binary forms, with or without
+#    modification, are permitted provided that the following conditions are
+#    met:
+#
+#    1. Redistributions of source code must retain the above copyright notice,
+#       this list of conditions and the following disclaimer.
+#
+#    2. Redistributions in binary form must reproduce the above copyright
+#       notice, this list of conditions and the following disclaimer in the
+#       documentation and/or other materials provided with the distribution.
+#
+#    3. Neither the name of the QuTiP: Quantum Toolbox in Python nor the names
+#       of its contributors may be used to endorse or promote products derived
+#       from this software without specific prior written permission.
+#
+#    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+#    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+#    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+#    PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+#    HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+#    SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+#    LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+#    DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+#    THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+#    (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+#    OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+###############################################################################
+
+import pytest
+import functools
+import os
+import tempfile
+import numpy as np
+
+
+def _add_repeats_if_marked(metafunc):
+    """
+    If the metafunc is marked with the 'repeat' mark, then add the requisite
+    number of repeats via parametrisation.
+    """
+    marker = metafunc.definition.get_closest_marker('repeat')
+    if marker:
+        count = marker.args[0]
+        metafunc.fixturenames.append('_repeat_count')
+        metafunc.parametrize('_repeat_count',
+                             range(count),
+                             ids=["rep({})".format(x+1) for x in range(count)])
+
+
+def _skip_cython_tests_if_unavailable(item):
+    """
+    Skip the current test item if Cython is unavailable for import, or isn't a
+    high enough version.
+    """
+    if item.get_closest_marker("requires_cython"):
+        # importorskip rather than mark.skipif because this way we get pytest's
+        # version-handling semantics.
+        pytest.importorskip('Cython', minversion='0.14')
+
+
+@pytest.hookimpl(trylast=True)
+def pytest_generate_tests(metafunc):
+    _add_repeats_if_marked(metafunc)
+
+
+def pytest_runtest_setup(item):
```

Comment:
If I understand correctly, this applies to all tests, where `item` is the test function (what pytest called "Node")?

Reply:
Yeah exactly.  There's a [whole bunch of hook functions](https://docs.pytest.org/en/2.7.3/plugins.html#well-specified-hooks) that you can fill in if you like.  Each hook applies to every relevant item in its scope.  Here, `conftest.py` is a magic filename that acts at directory-scope.  We could also set this function with a file, or within a class, and it would only apply to tests in the same scope.  We do that a little bit in some files with `pytest_generate_tests`.

I set this hook up so it can be added to easily, but right now all it does is just skip tests marked with `requires_cython` if Cython isn't available.

Reply:
Also, while this is quite an old comment now, `item` refers to an object `pytest.Item`, which subclasses `pytest.Node`.